### PR TITLE
feat(types): restrict attributes typing

### PR DESCRIPTION
**Pull request by lohart13**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description Of Change

Updates the attributes typing for find queries so that model attributes can be selected/suggested by IDE's.

### CODE REVIEW

File: packages/core/src/model.d.ts [hunk 1]
```diff
@@ -838,15 +838,16 @@ export type Order = Fn | Col | Literal | OrderItem[];
  */
 export type ProjectionAlias = readonly [string | Literal | Fn | Col | Cast, string];
 
-export type FindAttributeOptions =
+export type FindAttributeOptions<TAttributes = any> =
   | Array<string | ProjectionAlias | Literal>
+  | Array<Extract<keyof TAttributes, string>>
```
Comment (ephys) [inline packages/core/src/model.d.ts:843]: This is a good change but I wonder if we could go even further by only allowing known attributes when using a string

```suggestion
  | Array<Extract<keyof TAttributes, string> | ProjectionAlias | Literal>
```

If the user needs to reference an unknown attribute, they can still use `col` or [`identifier`](https://github.com/sequelize/sequelize/pull/15598)

Comment (lohart13) [reply packages/core/src/model.d.ts]: I was wondering about this and have updated the PR to only known attributes for string values.

File: packages/core/test/unit/query-generator/select-query.test.ts [hunk 2]
```diff
@@ -491,8 +503,6 @@ Only named replacements (:name) are allowed in literal() because we cannot guara
         attributes: [
           // these used to have special escaping logic, now they're always escaped like any other strings. col, fn, and literal can be used for advanced logic.
           ['count(*)', 'count'],
-          '.*',
-          '*',
```
Comment (ephys) [inline packages/core/test/unit/query-generator/select-query.test.ts:495]: Those are important to keep as they test against a vulnerability that used to exist. I'd prefer to use `// @ts-expect-error` here

**Review by ephys — CHANGES_REQUESTED** [review]

File: packages/core/test/unit/query-generator/select-query.test.ts (hunk 2 above)
Comment (lohart13) [reply packages/core/test/unit/query-generator/select-query.test.ts:494]: I've readded this now

File: packages/core/test/integration/json.test.ts [hunk 3]
```diff
@@ -73,6 +73,9 @@ describe('JSON Querying', () => {
 
   const vars = beforeAll2(async () => {
     class User extends Model<InferAttributes<User>, InferCreationAttributes<User>> {
+      @Attribute(DataTypes.INTEGER)
+      @PrimaryKey
```
Comment (ephys) [inline packages/core/test/integration/json.test.ts:77]: by the way technically you only need to add the declare field, the decorators are not necessary as the property will still be auto-added

**Review by ephys — APPROVED** [review]

File: packages/core/test/integration/json.test.ts (hunk 3 above)
Comment (lohart13) [reply packages/core/test/integration/json.test.ts]: Ok, I've removed these now

**Review by ephys — APPROVED** [review]

**Pull request merged**